crates/matrix-sdk-common/src/linked_chunk/builder.rs

```diff
@@ -0,0 +1,458 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::HashMap, marker::PhantomData, ptr::NonNull};
+
+use super::{
+    Chunk, ChunkContent, ChunkIdentifier, ChunkIdentifierGenerator, Ends, LinkedChunk,
+    ObservableUpdates,
+};
+
+/// A temporary chunk in the [`LinkedChunkBuilder`], using [`ChunkIdentifier`]
+/// as the temporary links to the previous and next chunks.
+struct TemporaryChunk<Item, Gap> {
+    id: ChunkIdentifier,
+    previous: Option<ChunkIdentifier>,
+    next: Option<ChunkIdentifier>,
+    content: ChunkContent<Item, Gap>,
+}
+
+/// A data structure to rebuild a linked chunk from its raw representation.
+///
+/// A linked chunk can be rebuilt incrementally from its internal
+/// representation, with the chunks being added *in any order*, as long as they
+/// form a single connected component eventually (if they not, then the final
```

Comment:
:nerd_face: WELLLLLLLLLL, HACKSHTUALLEYYYYYY :sweat_smile: :sweat_smile: :sweat_smile: since it stashes nodes with two unconstrained links, it's a graph *until* you call `build()`, which confirms if it was a linked list by not returning an error ^^ This is what all the checks there do, really; sounds more and more enticing to split them as a separate method/builder…